ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -211,6 +244,35 @@ public void readInto(ByteBuffer buffer, long position) throws IOException {
     }
   }
 
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * @param byteArray The buffer into which the data needs to be written.
+   * @param position The source's position to start the read.
+   * @param length The length of data in bytes to read.
+   * @throws IOException if data could not be written to the file because of I/O errors.
+   * @throws IndexOutOfBoundsException if {@code position} < header size or >= {@link #sizeInBytes()} or if
+   * {@code buffer} size is greater than the data available for read.
+   */
+  public void readIntoDirectly(byte[] byteArray, long position, int length) throws IOException {
+    long sizeInBytes = sizeInBytes();
+    if (position < startOffset || position >= sizeInBytes) {
+      throw new IndexOutOfBoundsException(
+          "Provided position [" + position + "] is out of bounds for the segment [" + file.getAbsolutePath()
+              + "] with size [" + sizeInBytes + "]");
+    }
+    if (position + length > sizeInBytes) {
+      metrics.overflowReadError.inc();
+      throw new IndexOutOfBoundsException(
+          "Cannot read from segment [" + file.getAbsolutePath() + "] from position [" + position + "] for size ["
+              + length + "] because it exceeds the size [" + sizeInBytes + "]");
+    }
+    DirectRandomAccessFile directFile = new DirectRandomAccessFile(file, "r", 2 * 1024 * 1024);
```

Comment:
do you need a new `DirectRandomAccessFile` instance each time, or is there a read method that can take in a file position like on line 243 with `FileChannel`.

If you do need to open one each time, use the try with resources syntax:
```java
try (DirectRandomAccessFile directFile = new DirectRandomAccessFile(file, "r", 2 * 1024 * 1024)) {
  directFile.seek(position);
  directFile.read(byteArray, 0, length);
}
```
This will ensure that `close()` will be called even if the read fails.

Reply:
`FileChannel` has the interface to read with a position, but `DirectRandomAccessFile` needs to `seek()` then `read()`. We need an extra `position` lock if a shared `DirectRandomAccessFile` is used. 

try-with-resouces looks good. Made the change.